exonum/src/runtime/mailbox/blockchain_secretary.rs

```diff
@@ -0,0 +1,321 @@
+
+                dispatcher.register_artifact(fork, &artifact, spec)?;
+
+                mailbox.add_notification(Notification::ArtifactDeployed { artifact });
+
+                Ok(())
+            }
+
+            // Mutable action.
+            Action::AddService {
+                artifact,
+                instance_name,
+                config,
+            } => {
+                // Request the dispatcher to start an instance of service given the
+                // deployed artifact.
+
+                let fork = fork.ok_or(DispatcherError::InappropriateTimeForAction)?;
+
+                let instance_spec = InstanceSpec {
+                    artifact,
+                    name: instance_name,
+                    id: dispatcher.assign_instance_id(fork),
+                };
+
+                dispatcher.add_service(fork, instance_spec.clone(), config)?;
+
+                let part_of_core_api = is_part_of_core_api(&instance_spec);
+                mailbox.add_notification(Notification::InstanceStarted {
+                    instance: instance_spec,
+                    part_of_core_api,
+                });
+
+                Ok(())
+            }
+
+            Action::UpdateConfig {
+                caller_instance_id,
+                changes,
+            } => {
+                // Request dispatcher to change the configuration of the blockchain
+                // part (internal or service).
+
+                let fork = fork.ok_or(DispatcherError::InappropriateTimeForAction)?;
+
+                dispatcher.update_config(mailbox, fork, caller_instance_id, changes);
+                Ok(())
+            }
+        })
+        .and_then(callback)
+    }
+
+    /// Checks if caller has sufficient rights to perform request.
+    fn verify_caller(&self, initiator: InstanceId) -> AuthoriziationResult {
+        let authorized_access = is_authorized_for_requests(initiator);
+
+        // In the transaction execution context unauthorized access is an error.
+        if let MailboxContext::TxExecution(ref call_info) = self.context {
+            if call_info.instance_id != initiator {
+                return AuthoriziationResult::AbortProcessing(
+                    DispatcherError::FakeInitiator.into(),
+                );
+            }
+
+            if !authorized_access {
+                return AuthoriziationResult::AbortProcessing(
+                    DispatcherError::UnauthorizedCaller.into(),
+                );
+            }
+        }
+
+        if authorized_access {
+            AuthoriziationResult::Valid
+        } else {
+            AuthoriziationResult::ShouldSkip
+        }
+    }
+}
+
+/// Internal function encapsulating the check for service
+/// to have sufficient rights to request actions from the blockchain.
+fn is_authorized_for_requests(instance_id: InstanceId) -> bool {
+    // Currently, only Supervisor service is authorized to request changes.
+    instance_id == crate::runtime::SUPERVISOR_INSTANCE_ID
```

Comment:
I think currently it's OK to have that check done in this way. However, since it's encapsulated, this check can be moved to the more appropriate place (e.g. `Blockchain`) easily.